bpo-30768: Recompute timeout on interrupted lock

vstinner opened the pull request:
Fix the pthread implementation of PyThread_acquire_lock_timed() when
called with timeout > 0 and intr_flag=0: recompute the timeout if
sem_timedwait() is interrupted by a signal (EINTR).

See also the PEP 475.

<!-- issue-number: bpo-30768 -->
https://bugs.python.org/issue30768
<!-- /issue-number -->

vstinner committed:
bpo-30768: Recompute timeout on interrupted lock

Fix the pthread+semaphore implementation of
PyThread_acquire_lock_timed() when called with timeout > 0 and
intr_flag=0: recompute the timeout if sem_timedwait() is interrupted
by a signal (EINTR).

See also the PEP 475.

pitrou commented on Python/thread_pthread.h:
```diff
@@ -318,23 +318,59 @@ PyThread_acquire_lock_timed(PyThread_type_lock lock, PY_TIMEOUT_T microseconds,
-        if (microseconds > 0)
+
+        if (!intr_flag) {
+            _PyTime_t timeout = _PyTime_FromNanoseconds(microseconds * 1000);
```
What if `microseconds * 1000` overflows?

vstinner replied:
I added _PyTime_FromMicroseconds() to detect overflow on "us * 1000". But I chose to not detect overflow on "_PyTime_GetMonotonicClock() + timeout".

pitrou commented on Python/pytime.c:
```diff
@@ -262,6 +262,20 @@ _PyTime_FromNanoseconds(long long ns)
+    _PyTime_t t = (_PyTime_t)us;
+
+    if (_PyTime_check_mul_overflow(t, US_TO_NS)) {
+        _PyTime_overflow();
```
I don't think it will be safe to raise an exception from a PyThread function (which could be called without the GIL).

vstinner replied:
Oops, you're right. I forgot that.

pitrou commented on Python/pytime.c:
```diff
@@ -301,9 +315,16 @@ _PyTime_FromTimeval(_PyTime_t *tp, struct timeval *tv, int raise)
 
-    t += (_PyTime_t)tv->tv_usec * US_TO_NS;
+    if (_PyTime_FromMicroseconds(&us, tv->tv_usec) < 0) {
+        if (!raise) {
```
Isn't it too late? You've already raised the exception.

vstinner replied:
I removed _PyTime_FromMicroseconds(), it was a bad idea.

vstinner committed:
Adjust PY_TIMEOUT_MAX to prevent overflow

vstinner commented:
I read again the code. PyThread_acquire_lock_timed() doc is explicit about the timeout limit: "microseconds **must be less than PY_TIMEOUT_MAX**. Behaviour otherwise is undefined."

I removed _PyTime_FromMicroseconds(), it was a bad idea. Instead, I ajusted PY_TIMEOUT_MAX.

By the way, the main consumer of PyThread_acquire_lock_timed() is threading.Lock.acquire() and this function is careful on integer overflow. This function already stores internally the timeout as a _PyTime_t and explicitely ensures that "microseconds <= PY_TIMEOUT_MAX".

vstinner commented:
My patch doesn't change the value of threading.TIMEOUT_MAX, it's still a value larger than 292 years :-)

```
>>> import threading
>>> threading.TIMEOUT_MAX
9223372036.0
>>> (2**63-1) // 10**9
9223372036
>>> threading.TIMEOUT_MAX / (3600 * 24 * 365.25) # years
292.27102301822697
```

If you ignore PyThread_acquire_lock_timed() doc and pass a timeout larger than 292 years, you get an overflow. Sorry for you. But why do you use such crazy timeout? :-)

vstinner committed:
Fix _thread.TIMEOUT_MAX rounding on Windows

vstinner committed:
Fatal error if timeout > PY_TIMEOUT_MAX

The pthread implementation of PyThread_acquire_lock() now fails with
a fatal error if the timeout is larger than PY_TIMEOUT_MAX, as done
in the Windows implementation.

The check prevents any risk of overflow in PyThread_acquire_lock().

Add also PY_DWORD_MAX constant.

vstinner commented on Python/thread_pthread.h:
```diff
@@ -318,23 +318,62 @@ PyThread_acquire_lock_timed(PyThread_type_lock lock, PY_TIMEOUT_T microseconds,
+            }
+            else if (dt > 0) {
+                _PyTime_t realtime_deadline = _PyTime_GetSystemClock() + dt;
+                if (_PyTime_AsTimespec(realtime_deadline, &ts) < 0) {
```
Oh, this function raises an exception on overflow. This bug is now fixed in the new commit.

pitrou replied:
Which new commit? :-)

vstinner replied:
4th commit of this PR, commit 7c428b7014ecd9058f06d30ccd4deb93a6c28c0e which added the following code at the top of the function:

```
+    if (microseconds > PY_TIMEOUT_MAX) {
+        Py_FatalError("Timeout larger than PY_TIMEOUT_MAX");
+    }
```

pitrou replied:
Ok, thank you for adding the comments :-)

pitrou commented:
> My patch doesn't change the value of threading.TIMEOUT_MAX, it's still a value larger than 292 years :-)

I see, thank you. I guess it's large enough :-)

pitrou commented on Include/pythread.h:
```diff
@@ -42,16 +42,23 @@ PyAPI_FUNC(int) PyThread_acquire_lock(PyThread_type_lock, int);
+#  define PY_TIMEOUT_MAX (PY_LLONG_MAX / 1000)
+#elif defined (NT_THREADS)
+   /* In the NT API, the timeout is a DWORD and is expressed in milliseconds */
+#  if 0xFFFFFFFFLL * 1000 < PY_LLONG_MAX
```
Which integer width does the C preprocessor use? :-) Can `0xFFFFFFFFLL * 1000` be truncated?

vstinner replied:
I didn't write this code, I only moved it.

The "LL" suffix is for "long long" which should be at least 64-bit signed integer, no?

Are you suggesting to replace the code with "0xFFFFFFFFLL < PY_LLONG_MAX / 1000" to prevent any risk of integer overflow?

pitrou replied:
Ah, ok. No, nevermind.

vstinner committed:
Remove now unused "exit" label

vstinner committed:
Update generated code: run "make clinic"

vstinner committed:
Merge branch 'master' into sem_timedwait

vstinner commented:
Oh, I forgot to regenerate generated file using "make clinic" when I added the PY_DWORD_MAX constant. It's now fixed. I also merged master into this PR.

pitrou commented:
Hmm... I *think* this PR looks ok now.

vstinner commented:
I tested manually the PR using files attached to the issue: "Apply PR 4103, apply attached test.patch, recompile Python, and run interrupted_lock.py to test the PR." It works well. Result:

```
haypo@selma$ ./python interrupted_lock.py 
acquire(timeout=1.0) took 1.073 seconds and got 105 signals
```

I modified interrupted_lock.py with:

```
def burn_cpu():
    for i in range(10**4 * 3):
        pass
```
